Introduce API endpoints for SEV attestation

vasiliy-ul opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:

The PR introduces new API endpoints for SEV attestation. It is based on the initial [design proposal](https://github.com/kubevirt/community/blob/main/design-proposals/amd-sev.md). It allows to perform the necessary pre-launch attestation steps:

- retrieving platform certificates for verification `sev/fetchcertchain`: ~the API call is routed to the node where the VM is scheduled and it is processed by `virt-handler`. The certificates are fetched from `libvirt` by `node-labeller` in the `init-container` context (thus ATM the support of certificates rotation is limited);~  UPD: the call is routed to the `virt-launcher` pod where the corresponding `libvirt` API is invoked;
- setting session parameters `sev/setupsession`: the call is routed to the `virt-launcher` pod where the corresponding `libvirt` API is invoked;
- querying launch measurement `sev/querylaunchmeasurement`: the call is routed to the `virt-launcher` pod where the corresponding `libvirt` API is invoked. The measurement is returned to the caller;
- injecting launch secret `sev/injectsecret`: the call is routed to the `virt-launcher` pod where the corresponding `libvirt` API is invoked.

The attestation process can be run either manually by calling the corresponding endpoints (e.g. using `virtctl`) or in an automated way by leveraging an attestation server and a controller that will monitor the VMI state and perform the required actions. The E2E functional test in `tests/launchsecurity/sev.go` demonstrates the general flow.

Attestation can be requested via VMI spec yaml:
```yaml
spec:
  startStrategy: "Paused"
  domain:
    firmware:
      bootloader:
        efi:
          secureBoot: false
    launchSecurity:
      sev:
        attestation: {}
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

- The implementation of the new `virtctl` commands mapped to the SEV endpoints is not in the scope of this PR and will be added later in a follow-up (same applies to the attestation controller: not in the scope ATM). UPD: https://github.com/kubevirt/kubevirt/pull/7595
- The PR requires `libvirt >= 8.0.0`: https://github.com/kubevirt/kubevirt/pull/7277
- There is a proposal to generalize the attestation flow in libvirt: https://listman.redhat.com/archives/libvir-list/2022-February/msg00857.html. Any thoughts on that are welcome.


**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Experimantal support of SEV attestation via the new API endpoints
```

vasiliy-ul commented:
/retest

dhiller commented:
/retest

vasiliy-ul commented:
/retest-required

iholder-redhat reviewed (commented):
Thanks @vasiliy-ul!
Thrilled to see this progressing!

I left some minor notes but overall I'm still digesting this PR :)
Thanks a lot

vasiliy-ul commented on pkg/virt-api/api.go:
```diff
@@ -458,6 +458,17 @@ func (app *virtAPIApp) composeSubresources() {
 			Returns(http.StatusOK, "OK", "").
 			Returns(http.StatusBadRequest, httpStatusBadRequestMessage, ""))
 
+		// AMD SEV endpoints
```
I added the comment just to improve the readability with my syntax highlight settings. It emphasizes the begining of the block with SEV endpoints. Without it all the endpoints from a quick look appear like identical blocks and indeed you need to read specifically the code to distinguish them.

I have this comment shown in bright yellow color on a dark background and it catches the attention immediately. Also this file already uses similar comments like `// VMI endpoint` or `// VM endpoint` so I do not see it as an issue to have `// AMD SEV endpoints` as well. But ofc, this is purely personal preference and if there are concerns I can remove it.

vasiliy-ul commented on pkg/virt-api/rest/subresource.go:
```diff
@@ -1259,3 +1263,154 @@ func (app *SubresourceAPIApp) VMIAddVolumeRequestHandler(request *restful.Reques
+	}
+
+	validate := func(vmi *v1.VirtualMachineInstance) *errors.StatusError {
+		if !vmi.IsScheduled() && !vmi.IsRunning() {
```
> Can you explain why !vmi.IsScheduled() part is essential?

In order to fetch the certificates we only need to know the correct worker node. The VMI in that case does not have to be in the `Running` state. It can still be `Scheduled`. This is in fact the case while it waits for the SEV session parameters fom the user.

> The error seems somewhat misleading that way

Yeah, agree. I will adjust it to reflect better the situation here. Thanks for pointing that.

alicefr replied:
@vasiliy-ul I think you need to remove the `!vmi.IsScheduled` right? If you use libvirt API you need to be able to talk to virt-launcher

vasiliy-ul replied:
No, this is a bit tricky. The VMI cannot transition to the `Running` state without the session parameters being provided. And a user can construct those only after fetching the SEV platform certs. So the flow is like the following:
- A VMI requesting SEV attestation is launched with empty `Session` and `DHCert` in yaml
- It gets **scheduled** to the proper node with SEV support
- `virt-handler` waits for the session parameters and meanwhile prevents the VMI from transitioning to the `Running` state
- User fetches the platform certs from the node
- User builds session parameters and calls `sev/setupsession` API
- The VMI spec gets patched with the session parameters
- `virt-handler` notices the session parameters and allows the transition to the `Running` state
- Then libvirt starts the domain in `Paused` state.

alicefr replied:
ok, but then you need to verify that the virt-launcher pod is running, otherwise you cannot talk to virt-launcher process, right?

alicefr replied:
I don't know it depends it can fail to fetch the platform certificates because virt-launcher hasn't been started yet. For example, I'm not sure if the VMI could be result is scheduled but virt-launcher hasn't been started yet because we are pulling a container image or because one of the init containers takes long

vasiliy-ul replied:
Looked through the code a bit. So here the `VMI` transitions to the `Scheduled` phase:

https://github.com/kubevirt/kubevirt/blob/a3aaf1ace40c44c9dafaf2c091851d5cdb829fd7/pkg/virt-controller/watch/vmi.go#L535

This code is inside the condition:

https://github.com/kubevirt/kubevirt/blob/a3aaf1ace40c44c9dafaf2c091851d5cdb829fd7/pkg/virt-controller/watch/vmi.go#L509

The function `IsPodReady` checks if the `virt-launcher` pod and the **compute container** are already running:

https://github.com/kubevirt/kubevirt/blob/a3aaf1ace40c44c9dafaf2c091851d5cdb829fd7/pkg/virt-controller/watch/vmi.go#L893-L915

Unless I fail to see something obvious that should work for the fetch-certs case. Anyway, if for whatever reason the pod is not ready, it will just return an error and the fetching can be retried.

vasiliy-ul commented:
> I left some minor notes but overall I'm still digesting this PR :) Thanks a lot

Thanks for the review, @iholder-redhat . I rebased the changes and addressed the comments.

iholder-redhat commented on tests/launchsecurity/sev.go:
```diff
@@ -1,27 +1,180 @@
+	)
+
+	newSEVFedora := func(opts ...libvmi.Option) *v1.VirtualMachineInstance {
+		const secureBoot = false
```
nit: maybe rename to "withoutSecureBoot"? or simply `libvmi.WithUefi(false)`, which is clear enough IMO

iholder-redhat commented on tests/launchsecurity/sev.go:
```diff
@@ -1,27 +1,180 @@
+			w.Write(uuid[8:])
+		}
+
+		// total length of table: header plus one entry with trailing \0
+		l := 16 + 4 + 16 + 4 + len(diskSecret) + 1
+		// SEV-ES requires rounding to 16
+		l = (l + 15) & ^15
```
This part is cryptic to me :)
Maybe it's better to introduce consts to explain these numbers' meaning?

iholder-redhat commented on tests/launchsecurity/sev.go:
```diff
@@ -38,4 +191,47 @@ var _ = Describe("[sig-compute]AMD Secure Encrypted Virtualization (SEV)", func(
+		measure := verifyMeasurement(&sevMeasurementInfo, tikBase64)
+		sevSecretOptions := encryptSecret(diskSecret, measure, tikBase64, tekBase64)
+
+		By("Injecting launch secret")
+		err = virtClient.VirtualMachineInstance(vmi.Namespace).SEVInjectLaunchSecret(vmi.Name, sevSecretOptions)
+		Expect(err).ToNot(HaveOccurred())
+
+		By("Unpausing the VirtualMachineInstance")
+		err = virtClient.VirtualMachineInstance(vmi.Namespace).Unpause(vmi.Name, &v1.UnpauseOptions{})
+		Expect(err).ToNot(HaveOccurred())
+		tests.WaitForVMIConditionRemovedOrFalse(virtClient, vmi, v1.VirtualMachineInstancePaused, 30)
```
One thing that I don't quite understand here:
As far as I can tell the VMI doesn't contain any encrypted disks that require an injected secret in order to boot. Is this correct? If it is, how this test validates that the injection process is successful?

Please correct me if I'm missing something here

iholder-redhat replied:
cool :) thanks

iholder-redhat reviewed (commented):
Another review cycle. Overall looks awesome! Thanks a lot!

iholder-redhat commented:
> Thanks for the review, @iholder-redhat.

No problem! :) I've reviewed again and left some comments. Thanks a lot for your effort!

> I rebased the changes and addressed the comments. Some things in this PR could be done differently. So I am open to discuss other approaches if any doubts ;)

To which other approaches do you refer? :smiley: Is it revolving around node-labeller implementation? Maybe it's good to open your concerns to discussions :)

vasiliy-ul commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -1751,6 +1756,102 @@ func (l *LibvirtDomainManager) GetFilesystems() ([]v1.VirtualMachineInstanceFile
+		return nil, err
+	}
+
+	sevMeasurementInfo.LoaderSHA = fmt.Sprintf("%x", h.Sum(nil))
```
> So IIUC, above (domainLaunchSecurityParameters.SEVMeasurement) is the state measurement from SEV hardware.

Correct. The measurement includes the initial memory state of the guest.

> Here we add another measurement that is the loader's 256-SHA.

This is not exactly 'another measurement'. This is the SHA digest of the OVMF binary `/usr/share/OVMF/OVMF_CODE.cc.fd` (which is included in the initial memory measuremnet mentioned above). The VM user will need this digest to calculate the measurement locally and then compare with the value returned from SEV hardware.

> So this will be the place to add more custom measurements in the future, right?

Well, in case when the VM is run with external kernel/initrd, the user will also require the appropriate digests of all the compnents: kernel, initrd, cmdline. But this is currently not implemented in this PR. The support of running an SEV VM with external kernel is still in TODO and will be added in a follow-up.

iholder-redhat replied:
Thanks for the clarifications!

vasiliy-ul commented on pkg/virt-api/rest/subresource.go:
```diff
@@ -1259,3 +1263,154 @@ func (app *SubresourceAPIApp) VMIAddVolumeRequestHandler(request *restful.Reques
 }
+
+func (app *SubresourceAPIApp) SEVFetchCertChainRequestHandler(request *restful.Request, response *restful.Response) {
+	if !app.clusterConfig.WorkloadEncryptionSEVEnabled() {
+		writeError(errors.NewBadRequest(fmt.Sprintf(featureGateDisabledErrFmt, virtconfig.WorkloadEncryptionSEV)), response)
+		return
+	}
```
I personally do not like putting error-checking code in helper functions like that. In my view this code is already clear enough. Do you think it would be better to have smth like bellow instead?
```go
if writeErrToResponseIfSevDisabled(response) {
    return
}
```

iholder-redhat replied:
Looks great :) thanks!

vasiliy-ul commented on pkg/virt-api/rest/vnc.go:
```diff
@@ -34,5 +34,8 @@ func validateVMIForVNC(vmi *v1.VirtualMachineInstance) *errors.StatusError {
 		log.Log.Object(vmi).Reason(err).Error("Can't establish VNC connection.")
 		return errors.NewBadRequest(err.Error())
 	}
+	if !vmi.IsRunning() {
+		return errors.NewBadRequest(vmiNotRunning)
+	}
```
> If someone tries to connect to a paused VMI's console currently, what happens?
You will see
`Successfully connected to sle15sp4-minimal console. The escape sequence is ^]`
but it will not accept any commands.

iholder-redhat replied:
In the PR you referred, IIUC, we're talking about an automated process revolving a console connection to a VMI that fails because of timeout. This is however is different than a user (non-automated) experience.

For example, as a user, isn't it reasonable to start a VM then right away try to connect to its console? Would users prefer to wait for a few seconds until the VM starts then having an error popped up?

vasiliy-ul replied:
> For example, as a user, isn't it reasonable to start a VM then right away try to connect to its console? Would users prefer to wait for a few seconds until the VM starts then having an error popped up?

I think currently with KubeVirt you cannot connect to the console untill a VM is in `Running` state. Just to note: this does not change the existing behaviour. I simply had to move this check as it previously was in 

https://github.com/kubevirt/kubevirt/blob/a3aaf1ace40c44c9dafaf2c091851d5cdb829fd7/pkg/virt-api/rest/subresource.go#L172-L177

In this PR I changed the condition: 

```go
func (app *SubresourceAPIApp) getVirtHandlerConnForVMI(vmi *v1.VirtualMachineInstance) (kubecli.VirtHandlerConn, error) {
	if !vmi.IsRunning() && !vmi.IsScheduled() {
		return nil, goerror.New(fmt.Sprintf("Unable to connect to VirtualMachineInstance because phase is %s instead of %s or %s", vmi.Status.Phase, v1.Running, v1.Scheduled))
	}
	return kubecli.NewVirtHandlerClient(app.virtCli).Port(app.consoleServerPort).ForNode(vmi.Status.NodeName), nil
}
```
Because the certificates should be fetched before a VM is in `Running`. But then I had to move this `IsRunning` check specifically to vnc and console

vasiliy-ul commented on cmd/virt-handler/virt-handler.go:
```diff
@@ -308,12 +308,16 @@ func (app *virtHandlerApp) Run() {
 			panic(err)
 		}
 		capabilities = nodeLabellerController.HostCapabilities()
+		pdh = nodeLabellerController.SEV.PDH
+		certChain = nodeLabellerController.SEV.CertChain
```
> Node labeller's job is to label the nodes. IIUC, in your implementation, this is not what's being done. Instead, the node labeller simply fetches the information from the node and hold it in-memory as part of its struct. This is misleading and not cohesive, that is, it's outside of node-labeller's responsibilities.


Currently node-labeller fetches the info and adds `kubevirt.io/sev` label to the node (which is later used for proper scheduling of a SEV VM):
https://github.com/kubevirt/kubevirt/blob/86f6b7daddeef6ad704c5d190bedf3c52ec4ebf7/pkg/virt-handler/node-labeller/node_labeller.go#L278-L282
Node-labeller also stores the capabilities. The fetched info includes the certificates and those can be considered as part of platform capabilities.

> virt-handler seems to me like a perfect candidate to fetch this data once needed. This way it wouldn't need to reach node-labeller. The initial implementation can be similar to what you have now, that is the handler can simply fetch is by parsing virsh nodesevinfo. In the future, this can also help solving the CA rotation problem you've mentioned in PR description.

Virt-handler cannot call directly `virsh nodesevinfo` since this needs to be done in the `virt-launcher` context. Initially I implemented it differently. Refering to the commented code https://github.com/kubevirt/kubevirt/pull/7197#discussion_r867452483: instead of `virsh nodesevinfo` I was fetching the certificates from `virt-launcher` by calling libvirt API `GetSEVInfo()`.

I was not sure which approach would be better here. I considered that the certs ware sort of node assets and they do not belong to a VM. Therefore I moved the fetching to node-labeller. So this was one point where I had doubts.

vasiliy-ul replied:
But now after some meditation I think the initial approach was right. The benefit of calling libvirt API directly on the VM is that indeed it automatically solves the problem with rotation of the certificates (since we will fetch the latest ones). I will need to double check that approach again but now I am in favour of restoring it instead of using the node labeller.

alicefr replied:
@vasiliy-ul do you need this information before scheduling the VM? Otherwise, you are starting the virt-launcher pod as paused, so you should have communication channel with virt-launcher already open. Or am I missing something?

vasiliy-ul replied:
@alicefr , right, the communication channel with `virt-launcher` is already available. In such case there is no need to call `virsh` since its possible to call [libvirt API](https://libvirt.org/html/libvirt-libvirt-host.html#virNodeGetSEVInfo) directly from `virt-launcher`. This approach also works fine. So there are just two options here: we can fetch the certs either from `node-labeller` of from `virt-launcher`.

With node-labeller I wanted to spare on  handler <-> launcher communication and also to cache the certs on the node so they could potentially be fetched without even having a VMI. Nevertheless I think I will switch back to the approach with calling the API in `virt-launcher`. There are some potential issues with permissinos but seems I managed to solve those. So it should work fine and this will automatically handle the certs rotation which is a nice benefit.

vasiliy-ul replied:
> the major change I think if you rely on libvirt API is that you need the VMi already up and running.

In fact no, `virNodeGetSEVInfo` does not require a running domain. It needs only a connection to libvirt which is available at that time.

> Why did you change the first approach?

Mostly because virt-launcher is not needed to query the certs since they are per-node assets. By using the node-labeller I was mostly trying to reduce the communication chain and simplify the implementation.

vasiliy-ul commented:
> 
> To which other approaches do you refer? smiley Is it revolving around node-labeller implementation? Maybe it's good to open your concerns to discussions :)

Yeah, sorry if my pev comment was not very specific. Correct, one of the doubdts I had was about using the node-labeller: https://github.com/kubevirt/kubevirt/pull/7197#discussion_r867469420

vasiliy-ul commented on pkg/virt-handler/rest/lifecycle.go:
```diff
@@ -240,3 +244,80 @@ func (lh *LifecycleHandler) getVMILauncherClient(request *restful.Request, respo
 }
+
+func (lh *LifecycleHandler) SEVFetchCertChainHandler(request *restful.Request, response *restful.Response) {
+	/*vmi, client, err := lh.getVMILauncherClient(request, response)
+	if err != nil {
+		return
+	}
+
+	log.Log.Object(vmi).Infof("Retreiving SEV platform info")
+
+	sevPlatformInfo, err := client.GetSEVInfo()
+	if err != nil {
+		log.Log.Object(vmi).Reason(err).Error("Failed to get SEV platform info")
+		response.WriteError(http.StatusInternalServerError, err)
+		return
+	}*/
```
yeah, that was the initial implementation for fetching hte certs. Related to the discussion in https://github.com/kubevirt/kubevirt/pull/7197#discussion_r867469420.

alicefr commented:
Overall looks good to me!

I just have a couple of questions/remarks:
1) Have you thought to unpause the VM with the secret injection? I totally understand if you want to keep separate the commands but it could avoid an additional interaction
2) Does the session string require special formatting? I have only see that you verify if it is empty or not.

vasiliy-ul commented:
> Overall looks good to me!
> 

Thanks for the review!

> I just have a couple of questions/remarks:
> 
> 1. Have you thought to unpause the VM with the secret injection? I totally understand if you want to keep separate the commands but it could avoid an additional interaction

Yes, I thought about it. In my opinion for an API it is better to follow the 'single responsibility' principle as it gives more flexibility and simplifies the implementation. The additional interaction can always be hidden from the end-user at a different level. E.g. in `virtctl sev inject-secret <VMI> --unpause`.

> 2. Does the session string require special formatting? I have only see that you verify if it is empty or not.

The session itself is a binary blob. The string is just base64 representation. I was not sure how to properly validate it.

vasiliy-ul commented:
Updated the PR. The certificates are now fetched in `virt-launcher` context. Node labeller is not used anymore. The modification is added as a separate commit.

iholder-redhat commented:
@vasiliy-ul Thanks very much for you hard work and patience :) Great stuff!
/lgtm
/reset-required

iholder-redhat commented:
sorry, typo.. :)
/retest-required

alicefr commented:
/lgtm

alicefr commented:
I didn't wanted to steal you assign :)

vasiliy-ul commented:
Rebased the PR. Resolved conflicts with the generated files. No functional changes.

vasiliy-ul commented:
/retest

alicefr commented on pkg/virt-api/rest/subresource.go:
```diff
@@ -1554,3 +1554,84 @@ func (app *SubresourceAPIApp) SEVQueryLaunchMeasurementHandler(request *restful.
+	}
+
+	opts := &v1.SEVSessionOptions{}
+	if request.Request.Body != nil {
```
Can we simplify here by removing the else  evaluting if the Body == nil before
```golang
if request.Request.Body == nil {
    writeError(errors.NewBadRequest(fmt.Sprintf(unmarshalRequestErrFmt, err)), response)
    return
}
err := yaml.NewYAMLOrJSONDecoder(request.Request.Body, 1024).Decode(opts)

```

alicefr commented on pkg/virt-api/rest/subresource.go:
```diff
@@ -1554,3 +1554,84 @@ func (app *SubresourceAPIApp) SEVQueryLaunchMeasurementHandler(request *restful.
+	if opts.Session == "" {
+		writeError(errors.NewBadRequest("Session blob is required"), response)
+		return
+	} else if opts.DHCert == "" {
```
The else is not needed as you are returning in the if

alicefr commented on pkg/virt-api/rest/subresource.go:
```diff
@@ -1554,3 +1554,84 @@ func (app *SubresourceAPIApp) SEVQueryLaunchMeasurementHandler(request *restful.
+		return
+	}
+
+	const patchFmt = `{ "op": "%s", "path": "/spec/domain/launchSecurity/sev", "value": %s }`
```
Can we use the functions in https://github.com/kubevirt/kubevirt/blob/main/pkg/util/types/patch.go instead of harcoding the patch

alicefr commented on pkg/virt-handler/rest/lifecycle.go:
```diff
@@ -240,3 +240,73 @@ func (lh *LifecycleHandler) getVMILauncherClient(request *restful.Request, respo
+	}
+
+	opts := &v1.SEVSecretOptions{}
+	if request.Request.Body != nil {
```
Same as the comment above you can evaluate if the Body == nil and return without the else

vasiliy-ul commented:
> Minor comments after another look but it looks great /lgtm

Thanks. All addressed.

iholder-redhat commented on staging/src/kubevirt.io/api/core/v1/types.go:
```diff
@@ -2331,3 +2331,49 @@ type PreferenceMatcher struct {
+// SEVPlatformInfo contains information about the AMD SEV features for the node.
+//
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type SEVPlatformInfo struct {
+	metav1.TypeMeta `json:",inline"`
+	// Base64 encoded platform Diffie-Hellman key.
+	PDH string `json:"pdh,omitempty"`
+	// Base64 encoded SEV certificate chain.
+	CertChain string `json:"certChain,omitempty"`
+}
```
I wonder why all of these types have to be defined here as these aren't API object IIUC. Is there a particular reason for that?

Also:
- Why is `metav1.TypeMeta` embedded here?
- Why is `deepCopy` being generated? IIUC it isn't being used anywhere. Further than that, copying such a struct that contains only primitives by-value seems straight forward, so why do we need this?

vasiliy-ul replied:
Right, these are not part of the VMI spec API. However they are part of the REST API. According to my observation `staging/src/kubevirt.io/api/core/v1/types.go` already includes structs not used in the VMI spec but used in REST calls (e.g. the Profiler types). Do you think better to move the SEV structs somewhere else?

Regarding `metav1.TypeMeta` and `deepCopy`: this is needed for the code generator:
```
make generate
...
# kubevirt.io/client-go/kubecli
../../vendor/kubevirt.io/client-go/kubecli/vmi.go:482:74: cannot use &sevPlatformInfo (type *"kubevirt.io/api/core/v1".SEVPlatformInfo) as type "k8s.io/apimachinery/pkg/runtime".Object in argument to v.restClient.Get().RequestURI(uri).Do(context.Background()).Into:
	*"kubevirt.io/api/core/v1".SEVPlatformInfo does not implement "k8s.io/apimachinery/pkg/runtime".Object (missing DeepCopyObject method)
make: *** [Makefile:42: generate] Error 1
```

iholder-redhat replied:
> Right, these are not part of the VMI spec API. However they are part of the REST API. According to my observation staging/src/kubevirt.io/api/core/v1/types.go already includes structs not used in the VMI spec but used in REST calls (e.g. the Profiler types). Do you think better to move the SEV structs somewhere else?

Well, not sure about it.
I guess we can keep it there unless @alicefr has another opinion

> Regarding metav1.TypeMeta and deepCopy: this is needed for the code generator:

:ok_hand:

alicefr replied:
mmh it is an interesting point. TBH, I think they fit in this file. One of the reasons why we don't have it in the VMI spec it's because it is a value set a runtime before launching the guest. This reminds me a bit  hotplugged volumes. We report the values in the status but we don't use the type in the VMI spec. Actually, I think it could be reasonable to register the SEVPlatformInfo value in the VMIStatus. @vasiliy-ul wdyt? It can be part of a later PR for me

vasiliy-ul replied:
Yep, initially I was thinking about having `SEVPlatformInfo` and `SEVMeasurementInfo` in the `VMIStatus`. This seems reasonable. Just decided to make things simple at first as this could be added in the follow-up PR.

alicefr replied:
sounds reasonable to me. @iholder-redhat ?

iholder-redhat replied:
Sounds great :+1: 
Thanks for the discussion

vasiliy-ul replied:
Added this to the tracking issue as a TODO item as a reminder.

iholder-redhat commented on tests/libvmi/vmi.go:
```diff
@@ -126,6 +126,14 @@ func WithSEV() Option {
 	}
 }
 
+func WithSEVAttestation() Option {
+	return func(vmi *kvirtv1.VirtualMachineInstance) {
+		startStrategy := v1.StartStrategyPaused
+		vmi.Spec.StartStrategy = &startStrategy
+		vmi.Spec.Domain.LaunchSecurity.SEV.Attestation = &v1.SEVAttestation{}
+	}
+}
+
```
I wonder if we need to also initiate `LaunchSecurity` here (as `WithSEV()` does). WDYT?

iholder-redhat reviewed (commented):
Thanks a lot @vasiliy-ul! Great work!
A few minor comments below

vasiliy-ul commented:
> Thanks a lot @vasiliy-ul! Great work! A few minor comments below

Thanks for the review. Comments addressed.

vasiliy-ul commented:
/retest

iholder-redhat commented:
Thanks a lot @vasiliy-ul! Great job!
/lgtm

vasiliy-ul commented:
/retest

vasiliy-ul commented:
/test pull-kubevirt-e2e-k8s-1.23-sig-storage-nonroot

vasiliy-ul commented:
Rebased and fixed the conflicts. No functional changes.

vasiliy-ul commented:
one more rebase...

alicefr commented:
/lgtm

xpivarc commented:
Hi @vasiliy-ul , great to see this effort of integrating the SEV into Kubevirt. I feel obligated to ask if there is a slight possibility to get this tested somehow. I think our CI is a great advantage of Kubevirt. Even periodic or release jobs on an internal node or external cluster would be sufficient IMHO.

vasiliy-ul commented:
Hi @xpivarc, this needs SEV-capable hardware for testing. AFAIK in KubeVirt CI there are no AMD machines, right? If so, then I am afraid testing becomes problematic.

xpivarc commented:
> Hi @xpivarc, this needs SEV-capable hardware for testing. AFAIK in KubeVirt CI there are no AMD machines, right? If so, then I am afraid e2e testing becomes problematic. What would you suggest?

You are correct. I was hoping we could add the needed hardware to our CI(similarly to how we have added SRIOV capable hardware to our cluster or how we got the ARM cluster). My point is that most approvers/maintainers don't have this hardware available and therefore it is very hard to verify and ensure this feature will be working. Do you think this would be feasible?

vasiliy-ul commented:
Valid point. This is why the feature is now stated as `experimantal`. Do you mean connecting the hardware from our test lab to KubeVirt CI? Well, the hardware I use for testing is available only internally from within our infra. Unfortunately, it is not accessible from the outside.

alicefr commented:
@vasiliy-ul do you think we might find a way to mock the libvirt golang calls? It isn't the best option for testing it but at least we validate the new KubeVirt API calls and flow

vasiliy-ul commented:
@alicefr, I don't have a solution on top of my hand right now, but I think yes, it's possible. Though, with mocked libvirt it won't be e2e testing anymore. I am personally not sure if it's worth the effort since the 'mocked' flow is supposed to be covered by the unit-tests.

alicefr commented:
@vasiliy-ul yes you are right :/ emulation of course also doesn't work. How do you run the tests, on kind? AFAIK, we cannot use nested, so this is an additional problem for kubevirtci

vasiliy-ul commented:
Yeah, there is no SEV emulation. Right, it also does not work nested :( I just run the tests on a physical AMD machine. Not using kind though, but other k8s distros (k3s and rke2 mostly).

rmohr commented:
> Valid point. This is why the feature is now stated as `experimantal`. Do you mean connecting the hardware from our test lab to KubeVirt CI? Well, the hardware I use for testing is available only internally from within our infra. Unfortunately, it is not accessible from the outside.

Maybe we can come up with a way of donating such a machine. One way would be sponsoring a bare-metal machine in IBM cloud. @fabiand FYI.

Maybe some the bare-metal machines are already AMD, or maybe we just have to rent a right one and we can use it with kind.
Let's see if we  can come up with a plan. I would prefer to have at least a rough discussion/idea before talking the PR in.

rmohr commented:
There seem to exist AMD offerings: https://cloud.ibm.com/gen1/infrastructure/provision/bm?packageId=2802&itemId=15824

rmohr commented:
@cwilkers @fabiand is there a way to donate resources or budget through CNCF for a project to have transparent donation reporting?

cwilkers commented:
> @cwilkers @fabiand is there a way to donate resources or budget through CNCF for a project to have transparent donation reporting?

I'm not sure, @aburdenthehand @jberkus  would you know?

aburdenthehand commented:
The CNCF has a [Community Infrastructure Lab](https://github.com/CNCF/cluster) that we can apply for. It's [hosted by Equinix ](https://metal.equinix.com/product/servers/) and they have two types of AMD machines available (C3 medium and M3 large). 
Would this be suitable?

alicefr commented:
We also need to consider that the host kernel requires the SEV support, so we need to check in which distro version is available

alicefr commented:
/remove-lifecycle stale

vasiliy-ul commented:
/remove-lifecycle stale

vasiliy-ul commented:
/test pull-kubevirt-e2e-k8s-1.26-sev

vasiliy-ul commented:
/test pull-kubevirt-e2e-k8s-1.26-sev

vasiliy-ul commented:
/test pull-kubevirt-e2e-k8s-1.23-sig-compute

vasiliy-ul commented:
Hey, just a gentle ping here. This PR is again ready for review. The new SEV test job is green:

https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/7197/pull-kubevirt-e2e-k8s-1.26-sev/1626499438073614336

What's changed:
- the code has been rebased on the latest main
- adapted the e2e tests to not use hardcoded hardware-specific SEV parameters
- introduced `SEV` label to filter the tests properly
- fixed issues with unit-tests on aarch64

The new code has been added in separate commits on top of the existing ones (to simplify the review process). If/When have some time, please take a look. Thanks!

vasiliy-ul commented:
/test pull-kubevirt-e2e-k8s-1.26-sev
/test pull-kubevirt-e2e-k8s-1.26-sig-storage 
/test pull-kubevirt-e2e-arm64

vasiliy-ul commented:
/test pull-kubevirt-e2e-k8s-1.26-sev

vasiliy-ul commented:
/test pull-kubevirt-e2e-k8s-1.26-sev

alicefr commented:
> Hey, @xpivarc, @rmohr, @alicefr, @iholder101, a gentle ping here :)
> 
> Firstly, thank you all for the previous reviews and comments. Now the topic about testing has been addressed. We currently have `pull-kubevirt-e2e-k8s-1.26-sev` job to test PRs on AMD hardware with SEV. There is also a periodic `periodic-kubevirt-e2e-k8s-1.26-sev`. This PR is currently green: https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/7197/pull-kubevirt-e2e-k8s-1.26-sev/1639203141775790080. How can we move this forward? Could you have one more look here? Let's discuss if there are other concerns.

hi @vasiliy-ul, I'll give another pass to the PRs. Will be the CI job mandatory or optional?

vasiliy-ul commented:
> hi @vasiliy-ul, I'll give another pass to the PR. Will be the CI job mandatory or optional?

Hi @alicefr, the presubmit job is not set as mandatory as of now. We have only one AMD node, which is not yet configured for parallel runs. Potentially we can set up kind there, but we also need to consider that there is a hardware limitation on how many SEV VMs can be run at the same time (and ensure that this does not end up in a bottleneck for the CI). Smth to work on in a follow-up I think.

alicefr commented:
> > hi @vasiliy-ul, I'll give another pass to the PR. Will be the CI job mandatory or optional?
> 
> Hi @alicefr, the presubmit job is not set as mandatory as of now. We have only one AMD node, which is not yet configured for parallel runs. Potentially we can set up kind there, but we also need to consider that there is a hardware limitation on how many SEV VMs can be run at the same time (and ensure that this does not end up in a bottleneck for the CI). Smth to work on in a follow-up I think.

Yeah, that's a good point. IMO, we should also only run the SEV tests on that machine and avoid that other test VMs are scheduled there. We'll probably need more fine-grained scheduling for the tests.

vasiliy-ul commented:
> we should also only run the SEV tests on that machine and avoid that other test VMs are scheduled there

I think it is already done like that. Hm... Maybe @brianmcarey can confirm?

brianmcarey commented:
> > we should also only run the SEV tests on that machine and avoid that other test VMs are scheduled there
> 
> I think it is already done like that. Hm... Maybe @brianmcarey can confirm?

Yes currently only the SEV tests lanes are run against the SEV node - as optional lanes as you said.

vasiliy-ul commented:
Rebased the changes.

/test pull-kubevirt-e2e-k8s-1.26-sev

vasiliy-ul commented:
/test pull-kubevirt-e2e-k8s-1.26-sev

vasiliy-ul commented:
/test pull-kubevirt-code-lint

brianmcarey commented:
> /test pull-kubevirt-code-lint

The lint lane is currently broken - there is a fix PR in the queue though -https://github.com/kubevirt/kubevirt/pull/9515

alicefr commented on tests/launchsecurity/sev.go:
```diff
@@ -1,28 +1,251 @@
+
+	// The implementation is based on
+	//   https://blog.hansenpartnership.com/wp-uploads/2020/12/sevsecret.txt
+	encryptSecret := func(diskSecret string, measure []byte, tikBase64, tekBase64 string) *v1.SEVSecretOptions {
```
@vasiliy-ul can be this replaced by the [sevctl secret build](https://github.com/virtee/sevctl#secret-build)

vasiliy-ul replied:
That would probably make sense. I will take a look on how to use sevctl here.

alicefr replied:
> > but we could also use it in the node-labeller instead of relying on libvirt endpoints
> 
> For that, we would need to include sevctl into vit-launcher image. Not sure that an administrative tool should be part of the workload container.

Right, my hope is that one day we won't use virt-launcher image in the privileged pod, but this is another story. So, I agree.

tylerfanelli replied:
@vasiliy-ul Which version of sevctl is currently available?

vasiliy-ul replied:
@tylerfanelli, its `sevctl-0.3.2-1.el9.x86_64.rpm` from centos 9 stream.

alicefr replied:
@vasiliy-ul we can leave it as a todo

xpivarc commented:
> Hey, @xpivarc, @rmohr, @alicefr, @iholder101, a gentle ping here :)
> 
> Firstly, thank you all for the previous reviews and comments. Now the topic about testing has been addressed. We currently have `pull-kubevirt-e2e-k8s-1.26-sev` job to test PRs on AMD hardware with SEV. There is also a periodic `periodic-kubevirt-e2e-k8s-1.26-sev`. This PR is currently green: https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/7197/pull-kubevirt-e2e-k8s-1.26-sev/1639203141775790080. How can we move this forward? Could you have one more look here? Let's discuss if there are other concerns.

I moved this to my next week's schedule. Glad to see we moved with the testing.

brianmcarey commented:
/test pull-kubevirt-e2e-k8s-1.26-sev

vasiliy-ul commented:
A gentle ping here. This PR would still benefit from some reviews :)

/test pull-kubevirt-e2e-k8s-1.26-sev

alicefr commented:
/lgtm

@vasiliy-ul there is a unit test failing in the ci

vasiliy-ul commented:
> [<img alt="@kubevirt-bot" width="20" height="20" src="https://avatars.githubusercontent.com/u/25687724?s=40&amp;v=4"> ](/kubevirt-bot) ** pull-kubevirt-unit-test ** — Job succeeded.

@alicefr, from what I can see, the unit tests job is green. Or do I miss smth? There is a failed storage e2e test, but I dont think its related to this PR.

alicefr commented:
mmh I see the `pull-kubevirt-goveralls` failed

vasiliy-ul commented:
ah... I did not look at that job as it's not mandatory. Let's re-run them both. Previously there were no such failure...

/test pull-kubevirt-goveralls
/test pull-kubevirt-unit-test

vasiliy-ul commented:
The failed test uses 'real' timeouts/sleeps

https://github.com/kubevirt/kubevirt/blob/08f1b5c57d0cef851f4ba8122d493d10a9b2caa7/pkg/virt-handler/retry_manager_test.go#L59-L63

Depending on the CI load, this may fail. Does not seem related to this PR.

vasiliy-ul commented:
/test pull-kubevirt-e2e-k8s-1.26-sev

vasiliy-ul commented:
/test pull-kubevirt-e2e-k8s-1.26-sig-compute

vasiliy-ul commented:
Hi @alicefr, @iholder101. Previously, you reviewed the code here. Could you please re-lgtm/approve the PR (as you see fit)? I now rebased it and fixed all the conflicts (mostly with the generated code). There are no functional changes. The only thing is that I've moved the addition of `sevctl` tool to the test image into a separate PR (which is already merged https://github.com/kubevirt/kubevirt/pull/9892). Thanks!

/test pull-kubevirt-e2e-k8s-1.26-sev

vasiliy-ul commented:
/test pull-kubevirt-e2e-k8s-1.26-sev

vasiliy-ul commented:
/test pull-kubevirt-e2e-k8s-1.26-sig-storage

vasiliy-ul commented:
> Hi @alicefr, @iholder101. Previously, you reviewed the code here. Could you please re-lgtm/approve the PR (as you see fit)? I now rebased it and fixed all the conflicts (mostly with the generated code). There are no functional changes. The only thing is that I've moved the addition of `sevctl` tool to the test image into a separate PR (which is already merged #9892). Thanks!
> 

@alicefr, @iholder101, a gentle ping :) Would appreciate it if you could re-lgtm or approve the PR. Thanks!

alicefr commented on pkg/virt-launcher/virtwrap/manager_test.go:
```diff
@@ -1268,6 +1271,74 @@ var _ = Describe("Manager", func() {
+
+			manager, _ := NewLibvirtDomainManager(mockConn, testVirtShareDir, testEphemeralDiskDir, nil, ovmfDir, ephemeralDiskCreatorMock, metadataCache)
+			sevMeasurementInfo, err := manager.GetLaunchMeasurement(vmi)
+			if runtime.GOARCH == "amd64" {
```
Shouldn't we skip all the suite if this isn't amd64?

vasiliy-ul replied:
I am not sure about skipping, since it may hide potential issues. If the code is expected to error-out on non-amd64, IMHO, better to have a test in place to ensure it. In this particular case, `manager.GetLaunchMeasurement` fails due to missing ovmf code binary with sev support.

alicefr replied:
Ok, fine to me!

alicefr commented on tests/launchsecurity/sev.go:
```diff
@@ -197,6 +187,31 @@ var _ = Describe("[sig-compute]AMD Secure Encrypted Virtualization (SEV)", decor
 		}
 	}
 
+	parseVirshInfo := func(info string, expectedKeys []string) map[string]string {
```
This might be a good function to move to a shared files for utils. It can be done in a following PR

vasiliy-ul replied:
Agree, this is a good candidate for common utils. Though, on the other hand, I think it makes sense to move it when there are several users of that function. Otherwise, the utils tend to grow.

alicefr commented on tests/launchsecurity/sev.go:
```diff
@@ -197,6 +187,31 @@ var _ = Describe("[sig-compute]AMD Secure Encrypted Virtualization (SEV)", decor
+		return entries
+	}
+
+	toUint := func(s string) uint {
```
Same as above we could move it to a common shared file

alicefr commented:
/approve
/hold
Let's give some time to @iholder101 if he wants to have another look, but to me, it looks good
@vasiliy-ul great work and thanks for the perseverance!

vasiliy-ul commented:
Thanks a lot @alicefr for the review and valuable input!

iholder101 commented on pkg/virt-api/rest/subresource.go:
```diff
@@ -1622,3 +1623,29 @@ func (app *SubresourceAPIApp) SEVFetchCertChainRequestHandler(request *restful.R
+	}
+
+	validate := func(vmi *v1.VirtualMachineInstance) *errors.StatusError {
+		if !vmi.IsRunning() {
```
In other similar functions you have `!vmi.IsScheduled()` condition. Why is it different here?

iholder101 commented:
Sorry for the delay @vasiliy-ul!

Thanks very much for this PR! Appreciate your contributions a lot, I know it wasn't easy :)

/lgtm
/unhold

vasiliy-ul commented:
Thank you for the review, @iholder101!

/test pull-kubevirt-unit-test

dbasunag commented:
Looks like this PR broke https://github.com/openshift/openshift-restclient-python/blob/master/openshift/dynamic/discovery.py#L133, which expects VMInstancesSEVFetchCertChain/VMInstancesSEVQueryLaunchMeasurement/VMInstancesSEVSetupSession/VMInstancesSEVInjectLaunchSecret to be of the format '<a>/<b>' 

cc: @myakove
